### PR TITLE
[SLAPI-415] Refactor client to use OAuth flow

### PR DESCRIPTION
**Pull request by RuyiLi**
# Purpose

The current constructor requires the user to obtain a valid access token beforehand. This PR adds functionality to support the standard OAuth flow (get auth URL, provide code, then authorize with OAuth).

These changes are breaking since the constructor signature has been modified, so a major version change is required.

# Related Material

https://www.freshbooks.com/api/authenticating-with-oauth-2-0-on-the-new-freshbooks-api

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -65,13 +95,13 @@ Example API client call:
...
+	// Get the current user
+	const { data } = await client.users.me()
 
-    console.log(`Hello, ${data.id}`)
+	console.log(`Hello, ${data.id}`)
 } catch ({ code, message }) {
-    // Handle error if API call failed
-    console.error(`Error fetching user: ${code} - ${message}`)
+	// Handle error if API call failed
+	console.error(`Error fetching user: ${code} - ${message}`)
```
Comment (amcintosh) [inline README.md:104]: Could we move these back to spaces from tabs? The rendered page looks a little inconsistent with this change.

File: packages/api/__tests__/APIClient.test.ts [hunk 2]
```diff
@@ -1,26 +1,76 @@
...
 
+		test('Test client config errors, missing redirect URI', async () => {
+			const client = new APIClient(clientId, {
+				clientSecret: 'token',
+			})
+
+			try {
+				await client.getAuthRequestUrl()
+			} catch (err: any) {
+				expect(err).toBeInstanceOf(APIClientConfigError)
```
Comment (amcintosh) [inline packages/api/__tests__/APIClient.test.ts:30]: 👍

File: README.md (hunk 1 above)
Comment (RuyiLi) [reply README.md:104]: Oops; I think this was a consequence of prettier formatting things automatically. Should I create a .prettierignore and add the documentation files to it?

File: packages/api/package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "@freshbooks/api",
-	"version": "2.1.0",
+	"version": "3.0.0",
```
Comment (amcintosh) [inline packages/api/package.json:3]: There's a process to bump this at release time. Should be fine to leave as-is until then.

**Review by amcintosh — APPROVED** [review]

**Pull request merged**